src/Microsoft.ApplicationInsights/PreciseTimestamp.cs

```diff
@@ -0,0 +1,109 @@
+﻿namespace Microsoft.ApplicationInsights
+{
+    using System;
+    using System.Diagnostics.CodeAnalysis;
+#if NET45 || NET46
+    using System.Diagnostics;
+    using System.Threading;
+#endif
+
+    internal class PreciseTimestamp
+    {
+        private static readonly object Lck = new object();
+        private static PreciseTimestamp instance = null;
+
+#if NET45 || NET46
+        private static TimeSync timeSync;
+        private readonly Timer syncTimeUpdater;
+
+        private PreciseTimestamp()
+        {
+            this.syncTimeUpdater = InitializeSyncTimer();
+        }
+#endif
+
+        public static PreciseTimestamp Instance
+        {
+            get
+            {
+                if (instance == null)
+                {
+                    lock (Lck)
+                    {
+                        if (instance == null)
+                        {
+                            instance = new PreciseTimestamp();
+                        }
+                    }
+                }
+
+                return instance;
+            }
+        }
+
+        /// <summary>
+        /// Returns high resolution (1 DateTime tick) current UTC DateTime. 
+        /// </summary>
+        [SuppressMessage("Microsoft.Performance", "CA1822:MarkMembersAsStatic", Justification = "Enforcing instance fields initialization.")]
```

Comment:
This 'smells' wrong. If it accesses only static fields, it should be a static method. "Enforcing instance fields" sounds like sub-optimal class design -- I think I know where it comes from: You have fields that are 'beforefieldinit'.
Suggestions:
The whole class ought to be static. There is no need for an "Instance" property. The one instance field (syncTimeUpdater) ought to be static too. Add a static constructor to ensure that it's created on first access.

Reply:
I tried to explain my motivation above - static constructor would be worse for perf than a singleton in this case.

Reply:
I've never heard of static constructor being a perf problem. I searched and found a stackoverflow post from 2010 that was quickly debunked. Do you have more information?
I admit I've never studied it, but that's probably because it's never shown up in any perf trace I've seen.

There _are_ surprises with beforefieldinit fields NOT being created in release bits if they're not actually touched by the JIT compiler. That can certainly be avoided with an explicit static constructor. So, I'd recommend that if we can settle the perf issue.